# Midterm study Notebook

The problems in Part One are taken from a previous quiz.  Write the value of each expression in the space provided. A helpful reminder.  When applied to strings, '<' checks alphabetical order.  So

```
>>> 'a' < 'b'
True
>>> 'b' < 'a'
False
>>> 'a' < 'a'
False
```

When applied to strings '+' does concatenation:

```
>>> 'a' + 'x'
'ax'
```


## Part One

In [3]:
lender_rates = {'Fannie Mae': .05, 'Bank of America': .065, 'Quicken Loans': .045}
prime_interest_rate = .04
amount_above_prime = [lender_rates[x] - prime_interest_rate for x in lender_rates.keys()]
loan_amounts = [100000, 150000, 200000, 500000]
interest_due = [x * prime_interest_rate for x in loan_amounts]
prime_mortgages = [x <= 200000 and x > 100000 for x in loan_amounts]
prime_mortgages2 = [x for x in loan_amounts if x <= 200000 and x > 100000]
shoe_string = 'adidas_ultra_boost'

In [4]:
(lender_rates, prime_interest_rate, amount_above_prime, 
 loan_amounts, interest_due, prime_mortgages)

({'Fannie Mae': 0.05, 'Bank of America': 0.065, 'Quicken Loans': 0.045},
 0.04,
 [0.010000000000000002, 0.025, 0.0049999999999999975],
 [100000, 150000, 200000, 500000],
 [4000.0, 6000.0, 8000.0, 20000.0],
 [False, True, True, False])

In [5]:
letters = 'abcdefghijklmnopqrstuvwxyz'
vowels = 'aeiou'
word = 'facetious'
X = [l for l in letters if l not in vowels]
Y = [x for x in letters if x > 'd' and x < 'j']
W = [x for x in letters if x == 'd' or x == 'j' or x < 'q']
Z = [x > 'd' and x < 'j' for x in letters]
R = [x for x in word if x in vowels]
consonants = [l for l in letters if l not in vowels]
#. This is a dictionary comprehension.  It creates a dictionary
consonant2vowel = {consonants[i]:v for (i, v) in enumerate(vowels)}
vowel2consonant = {v:k for (k,v) in consonant2vowel.items()}

# Exercises

# A

Write some code to turn `X` into a string. Remember `X` was created by a list comprehension, so it is initially defined to be a list.  The string you should get is:

```
'bcdfghjklmnpqrstvwxyz'
```

Solution:

# B

Write some code to produce a list of letters that are in `letters` but are not in `Y`.  Your code should use the current values of `letters` and `Y` and it should be a list comprehension using a filter.  Your result should look like this:

```
['a', 'b', 'c', 'd', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z']
```

Solution:

# C

Write some code that returns a string consisting of the letters that are both in `W` and in `Y`.  

Solution:

# D

Write a function called `find_vowels` that will take any string and return a **string** containing the vowels in that string.  For the vowels use  `vowels` (defined above)  After writing the function, you should
be able to apply it to `word` (as defined above).  The answer would be

```
>>> find_vowels(word)
'aeiou'
```

Then you should be able to
apply the function to `word` again and see whether the function returns a value `==` to `vowels`.
The result you should get is:

```
>>> find_vowels(word) == vowels
True
```

Notice that this last test also verifies that your function
returns a string, because only a string can be == to
a string.

You should be able to apply the function to `'radish'`.  The answer should be:

```
>>> find_vowels("radish")
'ai'
```

You should able to concatenate the answer the last question with
'x' as follows:

```
>>> find_vowels("radish") + 'x'
'aix'
```

Notice that this last test is a second way of verifying that your function
returns a string, because only a string can be concatenated with
a string.

Solution:

Comments:  you lose points if you wrote a function that printed something but has no
`return` statement.   You lose points if your code handles two test cases instead of all
three, and even more points if it handles one out of three, and still more if it handles
none.

You lose points if you return the wrong type of Python object.  The directions above clearly 
asked for a string.  If you put square brackets around your string, you are returning
a list of strings, not a string.  If you put curly braces around your string you are returning
a set of strings, not a string.  And so on.  Some of the above tests were devised so as to make you
think about this. 

# E

Modify `find_vowels` as defined in problem  **D** so that its argument is optional and it behaves as follows:

```
>>> find_vowels() == vowels
True
```

Note:  All the previously discussed examples using `find_vowels` should continue to behave as before.

Solution:

# F 

Now write a function `same_chars` that
determines if two sequences have exactly the same
set of characters.  It should behave as follows:

```
>>> same_chars(vowels, 'aeiou')
True
>>> same_chars(vowels, 'eioua')
True
>>> same_chars(vowels, 'uoiea')
True
>>> same_chars(vowels, 'eiou')
False
>>> same_chars('bcd','dcb')
True
>>> same_chars('bcd','dcb') == True
True
>>> if same_chars('xyz','zyx'):
...     print('Same!`)
... else:
        print('Different')
Same!
```

The last two tests verify that you are returning `True` or `False`, not just printing them.
If you don't know what this means, study the functions nb and the functions assignment
solution.

Hint: Use sets. Remember how easy it is in Python to turn one type into another.  For example:

```
>>> list('abc')
['a', 'b', 'c']
```


Solution:

# G

Consider `R` as defined above. Remember `R` filters the elements of `word` with the condition `x in vowels`.

Write some code that performs a test to show that `vowels` (a string) contains the same 
characters as `R` (a list).

Solution:

# H

Consider `consonant2vowel`.  It is a mapping from consonants to vowels.  Write some code that returns the vowel associated with `"b"` in `consonant2vowel`.  Note:  This question has a **very** simple answer.  It is about
whether you understand what type of data structure `consonant2vowel` is, and how you use it.

Solution:

# I

Write a simple one-line Python expression that operates on `vowels` and returns 

```
['ba', 'be','bi', 'bo','bu']
```

Solution:

# J

Look at `interest_due`.  Using that as a model, write a simple expression that returns a list containing
the result of dividing every loan amount in `loan_amounts` by 100,000.


Solution:

# K

Write a simple one line Python expression that operates on X and returns

```
['b', 'c']
```

Use slices.

Solution:

# L


Write a Python expression which operates on `letters` and returns the following value. 

```
'ghi'
```

Use slices.

Solution:

# M

Write a Python expression which operates on `Z` and returns the following value

```
[False, False, False]
```

Solution:

# N

Write a Python expression which operates on `Y` and returns the following value:

```
['e','i']
```

Solution:

# O

Write a line of code which operates on `Y` and returns `False`.

Solution:

# P

Write a line of code which accesses `vowels` and returns the following value:

```
[True, True, False, False, False]
```

There are many solutions.  Just find one.

Solution:

# Q

Write a line of code which operates on `consonant2vowel` and returns the following value.

```
'e'
```

Solution:



Write a line of code which operates on `vowel2consonant` and returns the following value.

```
'g'
```

Solution:

##       Part Two

Write functions that behave according to the descriptions given.

# 1

Write a function that takes two numbers as arguments and multiplies them together and adds 1 to the result.  Call your function `product_successor`.  Then:

```
>>> product_successor(3,2)
7
>>> product_successor(5,9)
46
>>> product_successor(5,9) == 46
True
```

Last test verifies that the functioin is returning the value, not just printing it.

Solution:

# 2

Write a function called `product_successor_ratios`
which takes four numbers as arguments.   Call them `a`,`b`,`c`' and `d`;
`product_successor_ratios` should first find the product successor of `a` and `b`.

That is, it should multiply `a` and `b` together
and add one to the result.  It should call the function `product_successor` (defined in the last problem))
to do this.  Note: **No credit for a solution that ignores this instruction**.   

`product_successor_ratio` should
also call `product_successor` on `c` and `d` (multiplying `c` and `d` together and adding 1 to the result).
It should return the ratio obtained by dividing the product successor of `a` and `b` 
by the product successor of `c` and `d`.

Note: It took a lot of English to say this but the body of the `product_successor_ratios`
function is essentially one line of code.  Here are some examples illustratinmg its behavior:

```
>>> product_successor_ratio(3,2,5,9)   # Test One
0.15217391304347827
>>> product_successor_ratio(5,9,3,2)   # Test Two
6.571428571428571
>>> product_successor_ratio(3,2,5,9) + product_successor_ratio(5,9,3,2)   # Test Three
6.72360248447205
>>> product_successor_ratio(5,9,3,0)    # Test Four
46.0
>>> product_successor_ratio(5,9,0,0)    # Test Five
46.0
>>> product_successor_ratio(5,9,0,0) == 46.0    # Test Five
True
```


Finally,  the last two arguments of `product_successor_ratio` should be
optional.  Here are the results when they are omitted.

```
>>> product_successor_ratio(5,9,1)
23.0
>>> product_successor_ratio(5,9)
23.0
```


Solution:

## Part 3

Reminders: `range(n)` returns a list of the integers from 0 up to but not including `n`.

```
>>> list(range(10))
[0,1,2,3,4,5,6,7,8,9]
```

`'str0'.join(str_seq)`  returns a string gotten by joining together the elements of `str_seq` with `str0`. It is the opposite of `split`.  It is an error if the elements of `str_seq` are not all strings:

```
>>> ''.join(['a','b','c'])
'abc'
>>> 'x'.join(['a','b','c'])
'axbxc'
>>> 'x'.join([1,2,3])
TypeError      
...
TypeError: sequence item 0: expected string, int found
>>> 'i'.join(['m','ss','ss','pp',''])
'mississippi'
>>> 'i'.join(['m','ss','ss','pp'])
'mississipp'
```




In [55]:
letters = ('a', 'b', 'c', 'd', 'e', 'f', 'g', 
           'h', 'i', 'j', 'k', 'l', 'm', 'n', 
           'o', 'p', 'q', 'r', 's', 't', 'u', 
           'v', 'w', 'x', 'y', 'z')
vowels = 'aeiou'
digits = list(range(4))
word_list = ['bbop','pine', 'bake', 'isle']
weird_list = [word_list[x][x] for x in digits]
word3 = ''.join([x[1] for x in word_list])
word4 = 'x'.join(['o','o','o',''])
char_list = [x for x in word4 if x in vowels]
word1 = 'rambunctious'
word2 = 'mellifluous'
WW = word1 + word2
Z = [x for x in WW if x in 'lmn']
favorite_words = {'Noam Chomsky': word2, 'Leonard Bloomfield': word1, 'John Rupert Firth': weird_list,
                 'Ferdinand de Saussure': word3}

# a

Write out an expression which applies to `word3` and returns

```
'bs'
```

There is more than one solution.

Solution:

# b

Find a value for `LL` (a list)  for which the following piece of code produces the result shown.  There are some helpful examples above.

```
>>> 'a'.join(LL)
'abracadabra'
```

Solution:

Fill in your elements for `LL`:

In [2]:
LL=["????"]

# c

What is the Python type of `digits` (as defined above)?

Solution

[Write here]

# d

What is the Python type of `digits[0]`?

Solution:

[Write here]

# e

Find a value for `word_list2` (a list) which makes the expression below have the given value
(see the example using `weird_list` above).

```
>>> [word_list2[x][x] for x in digits]
['v ','o','t','e']
```


Solution:

Fill in your own elements for `word_list2` (there are many solutions).

In [3]:
word_list2 = ["????"]

# f

Write an expression which returns the **set** of vowels in  `char_list`.

There are several solutions. Find one.

Solution:

# g 

Write an expression which operates on `favorite_words` and returns:

```
'ias'
```

Hint: Use a slice.  Remember that `favorite_word` is a dictionary. Values are retrieved from a dictionary by keys, and therefore it makes no sense to use numerical indices (or slices) directly on a dictionary. 

Solution:

# h

Write an expression which operates on `favorite_words` and returns:

```
'bun'
```


 Solution:

# i

Write an expression which retrieves

```
['a','b','c','x','y''z']
```

from `letters`.  Hint: use a list comprehension with a filtering condition.  This is not slice territory.

Solution

# j

What is the effect of or the value returned by the following code?  Explain
by referring to the type of `letters`.

In [48]:
letters[1]  = '0'

Solution

[Write your answer here]

# k


What is the effect of or the value returned by the following code?  Explain
by referring to the type of `favorite_words`.

In [ ]:
favorite_words[0]  = 'Capitalism'

Solution:

[Write your answer here]

# l


What is the effect of or the value returned by the following code?  Explain
by referring to the type of `favorite_words`.

In [ ]:
favorite_words[2]

Solution:

[Write your answer here]

## Part 4

In [153]:
employees = ['happy', 'sleepy', 'dopey', 'bashful', 'grumpy', 'doc', 'sneezy']
employeeEmail = {'happy': ['sleepy','dopey'], 'dopey':['dopey'],'doc': employees,
                 'sleepy': ['bashful','doc', 'happy'],
                 'bashful': ['dopey']}
secret_word = 'utilitarian'
flip = {'0':'1', '1':'0'}
bit_string = '001110'
flipped = [flip[x] for x in bit_string]

# 1

Write an expression which operates on `employees` and returns

```
['dopey', 'bashful', 'grumpy']
```

Solution

# 2

Write an expression which operates on `employees` and returns

```
'y'
```

Solution:

# 3

Write an expression which operates on `employeeEmail` and returns

```
"rump"
```

Solution:

# 4

Write an expression which operates on `secret_word` and returns

```
'arian'
```

Solution:

# 5

Write an expression which operates on `secret_word` and returns

```
[False, False, True, False, True, False, True, False, True, True, False]
```

Solution:

# 6

`flipped` is defined above as

```
flipped = [flip[x] for x in bit_string]
```

The current value of `flipped` is

```
['1', '1', '0', '0', '0', '1']
```

Redefine `flip` so that the value of `flipped` becomes

```
[0, 0, 1, 1, 1, 0].
```


Solution:

Part 5

In [205]:
import numpy as np
x = np.array([5,13,23,47,9])
y = np.ones((5,))
z = np.ones((5,2))
w = np.arange(15).reshape((3,5))

In [206]:
w

array([[ 0,  1,  2,  3,  4],
       [ 5,  6,  7,  8,  9],
       [10, 11, 12, 13, 14]])

# a

Write out an expression which returns the shape of `x`,

Solution

# b

Write out an expression which operates on `x` and returns

```
array([10, 26, 46, 94, 18])
```

Note: A solution using a `for`-loop or a list comprehension will receive no credit.

Solution One:

# c

Write out an expression that returns an array containing the elementwise sum of the elements in `x` and `y`.

Solution:

# d

Explain why you get a `ValueError` for `x + z`. 

Solution:

[Write your answer here.]

# e

Note the following:

In [31]:
x < 20

array([ True,  True, False, False,  True], dtype=bool)

The result above is called a **Boolean array** because it only contains instances of the
type `bool` (`True` and `False`). Such Boolean arrays can be used to index arrays (basically
if they are of the same length).  Using a Boolean array this way is called using a **Boolean mask**.

# e (ctd.)

Using a Boolean mask, write an expression that returns an array containing the members of `x` that
are less than `20`.

Solution

# f

Now consider

In [196]:
two_d_array = np.array([[1,2],[2,1],[5,5],[9,4]])
two_d_array

array([[1, 2],
       [2, 1],
       [5, 5],
       [9, 4]])

# h.1 

Write an expression that returns the second row of `two_d_array`.

Solution:

# h.2 

Write an expression that returns `two_d_array` omitting the first row.

Solution:

# h.3

Write an expression that returns the following subarray of `two_d_array`.

```
array([[2, 1],
       [5, 5]])
```

Solution:

# i

Here is `w` again as a reminder. 

```
array([[ 0,  1,  2,  3,  4],
       [ 5,  6,  7,  8,  9],
       [10, 11, 12, 13, 14]])
```

Use fancy indexing to retrieve the following array from `w`

```
array([ 2,  6, 13])
```

Note:   You answer should be one Python expression, and it should use fancy indexing. not Boolean
indexing.

Solution

# j

Below we load the `iris` data discussed in lecture and review some features:

In [211]:
from sklearn.datasets import load_iris
data = load_iris()
features = data['data']
target = data['target']

In [212]:
features.shape

(150, 4)

Each row represents 4 measurements taken of an individual iris.

In [213]:
features[:10]

array([[5.1, 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2],
       [4.7, 3.2, 1.3, 0.2],
       [4.6, 3.1, 1.5, 0.2],
       [5. , 3.6, 1.4, 0.2],
       [5.4, 3.9, 1.7, 0.4],
       [4.6, 3.4, 1.4, 0.3],
       [5. , 3.4, 1.5, 0.2],
       [4.4, 2.9, 1.4, 0.2],
       [4.9, 3.1, 1.5, 0.1]])

Various kinds of infromation anbout the dataset are included in `data`, a dictionary-like object.

In [214]:
data.keys()

dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename', 'data_module'])

The measurements represented in the 4 columns are: 

In [215]:
data['feature_names']

['sepal length (cm)',
 'sepal width (cm)',
 'petal length (cm)',
 'petal width (cm)']

So for example the petal length of the first iris (row 0 printed above) is 1.4.

The three species we are trying to classify are named in target names.

In [216]:
data['target_names']

array(['setosa', 'versicolor', 'virginica'], dtype='<U10')

The array `target` (length 150) has the species for each iris in the data,
but represented as an integer.

In [217]:
set(target)

{0, 1, 2}

# j.1 

Your first task is to define a single Boolean array that is `True` for those rows of `features` that are not of target species `1`,
have sepal length greater than 4.7, and have petal width greater than 2.  Name your Boolean array `BC`.

Say what you think the length of  `BC` will be.

Solution:

[Write the value you expect for the length of `BC` here.]

# j.2

Your next task is to write an expression that computes how many rows there are
that are not of target species `1`,
have sepal length greater than 4.7, and have petal width greater than 2.

Your expression should use `BC`. In fact, there are 23 such rows, so the value of your expression
should be 23.

Solution:

# j.3

Your final task is to write an expression that returns the rows of `features` that are not of target species 1, have sepal length greater than 4.7, and have petal width greater than 2.  It should use `BC`.

Solution:

# j.4

Write an expression that computes how many of the rows computed in `j.3` have target species `0`. (In fact it's `0`,
so your expression will have `0` as its value).

Solution:

# j.5


Write another  expression that computes how many of the rows computed in `j.3` have target value `2`. (In fact it's 23,
so your expression will have `23` as its value).

Solution: